docs/extending/code-overview/evaluation.rst

```diff
@@ -75,7 +76,7 @@ expression that may not be visible. A later step may decide to
 materialize a value or convert the representation of ``Pi`` into a
 unicode string.
 
-And when an M-expression or an Expression doesn't match a particlar
+And when an :term:`M-expression` or an Expression doesn't match a particlar
```

Comment:
I believe that in the long term M-Expression, S-Expression and Expression should be the same thing.

Reply:
Here's the situation. S-Expression is what the majority of computer science-minded or programmers will know. M-Expression though happens to be the precise term. 

Expression, I think kind of means Mathics Expression, or an object of type Expression in the Mathics code. Right now we are supposed to think of it as a M-Expression even though it's API isn't that and who knows there may be side information squirreled away or hidden in here (which shouldn't be there). I would not even be surprised if _I_ were one of the people who contributed to this bad behavior.  But in an S-Expression and M-Expression this kind of thing shouldn't happen. 

In the future though as @mmatera's diagram has it (although that uses the word S-Expression) we have S-Expression gets converted to Expression which means we have more flexibility over the representation(s). It should probably be M-Expression compatible or convertible on demand. What goes in there is up to us and up for discussion. It probably _would_ represent associations as dictionaries for parts of the expression. 


 If it helps and it probably does we should try to be clarify what is meant by plain "Expression" and how that is different from M/S-expression.

Reply:
Regarding this, I think that what matters here is the interface that ``Expression``,  ``Symbol`` and ``Atom``defines. Then, you can have a ``String``, a ``MachineReal``, or an  ``Image``, and the evaluation algorithm will interpret them as part of an (S/M/)-Expression.